lib/src/switch.dart

```diff
@@ -61,20 +61,38 @@ extension SwitchLatest<T> on Stream<Stream<T>> {
       StreamSubscription<T>? innerSubscription;
       var outerStreamDone = false;
 
-      final outerSubscription = listen(
-          (innerStream) {
-            innerSubscription?.cancel();
-            innerSubscription = innerStream.listen(controller.add,
-                onError: controller.addError, onDone: () {
-              innerSubscription = null;
-              if (outerStreamDone) controller.close();
-            });
-          },
-          onError: controller.addError,
-          onDone: () {
-            outerStreamDone = true;
-            if (innerSubscription == null) controller.close();
+      void listenToInnerStream(Stream<T> innerStream) {
+        assert(innerSubscription == null);
+        var subscription = innerStream
+            .listen(controller.add, onError: controller.addError, onDone: () {
+          innerSubscription = null;
+          if (outerStreamDone) controller.close();
+        });
+        // If a pause happens during an innerSubscription.cancel,
+        // we still listen to the next stream when the cancel is done.
+        // Then we immediately pause it again here.
+        if (controller.isPaused) subscription.pause();
+        innerSubscription = subscription;
+      }
+
+      final outerSubscription =
+          listen(null, onError: controller.addError, onDone: () {
+        outerStreamDone = true;
+        if (innerSubscription == null) controller.close();
+      });
+      outerSubscription.onData((innerStream) {
+        var currentSubscription = innerSubscription;
+        if (currentSubscription != null) {
+          innerSubscription = null;
+          outerSubscription.pause();
+          currentSubscription.cancel().whenComplete(() {
```

Comment:
Thinking more about it I'm leaning towards forwarding errors to the result stream.

Reply:
That should be possible. I'd do something like:
```dart
Future<void>.sync(currentSubscription.cancel).catchError(controller.addError).whenComplete(() {
  ...
});
```
That should also catch synchronous errors from `cancel` (just to be paranoid).
I *think* it's safe to use `controller.addError` unconditionally. There is no chance that it has been closed if we paused `outerStream` when we received the previous stream event which caused the current cancel.